src/fmu/tools/qcforward/_qcforward_data.py

```diff
@@ -0,0 +1,309 @@
+
+        if "gridprops" in self._reuse:
+            CMN.print_info("Reuse current grid properties...")
+            return
+
+        CMN.print_info("Reading grid properties...")
+        if "gridprops" not in self._data.keys():
+            return
+
+        gprops = []
+        if self._project is None:
+
+            for mytuple in self._data["gridprops"]:
+                pname, pfile = mytuple
+
+                gridproppath = join(self._path, pfile)
+                current = xtgeo.gridproperty_from_file(
+                    gridproppath, name=pname, grid=self.grid
+                )
+                gprops.append(current)
+        else:
+            for pname in self._data["gridprops"]:
+                current = xtgeo.gridproperty_from_roxar(
+                    self._project, self._data["grid"], pname
+                )
+                gprops.append(current)
+
+        self._gridprops = xtgeo.GridProperties()
+        self._gridprops.append_props(gprops)
+
+    def read_wells(self):  # pylint: disable=too-many-branches
+        """Reading wells"""
+
+        if "wells" in self._reuse:
+            CMN.print_info("Reuse current wells...")
+            return
+
+        if "wells" not in self._data.keys():
+            return
+
+        CMN.print_info("Reading wells...")
+        wdata = []
+
+        # pylint: disable=too-many-nested-blocks
+        if self._project is None:
+            # fields may contain wildcards for "globbing"
+            if isinstance(self._data["wells"], list):
+                for welldata in self._data["wells"]:
+                    abswelldata = join(self._path, welldata)
+                    for wellentry in glob(abswelldata):
+                        mywell = xtgeo.Well()
+                        mywell.from_file(wellentry, lognames=self._wells_lognames)
+
+                        if self._wells_depthrange:
+                            tmin, tmax = self._wells_depthrange
+                            mywell.limit_tvd(tmin, tmax)
+
+                        if self._wells_rescale:
+                            mywell.rescale(self._wells_rescale)
+
+                        wdata.append(mywell)
+                        CMN.print_debug(wellentry)
+
+        else:
+
+            # roxar API input:
+            # wells: {"names": WELLS, "logrun": LOGRUN, "trajectory": TRAJ}
+
+            wnames = self._data["wells"].get("names", None)
+            wlogrun = self._data["wells"].get("logrun", "log")
+            wtraj = self._data["wells"].get("trajectory", "Drilled trajectory")
+
+            rmswells = [wll.name for wll in self._project.wells]
+            CMN.print_debug("All RMS wells: {}".format(rmswells))
+
+            CMN.print_debug("Data wells to match: {}".format(wnames))
+
+            for rmswell in rmswells:
+                for wreg in wnames:
+                    CMN.print_debug("Trying match {} vs re {}".format(rmswell, wreg))
+                    if re.match(wreg + "$", rmswell):
```

Comment:
It looks like a well will be loaded multiple times if it matches multiple regexes, maybe we should write something like:

```
if any(re.match(wreg + "$", rmswell) for wreg in wnames):
```

Reply:
Good point

Reply:
Will test in next PR